bench/wtperf/wtperf_throttle.c

```diff
@@ -0,0 +1,108 @@
+/*-
+ * Public Domain 2014-2015 MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "wtperf.h"
+
+/*
+ * Put the initial config together for running a throttled workload.
+ */
+int
+setup_throttle(CONFIG_THREAD *thread) {
+	THROTTLE_CONFIG *throttle_cfg;
+	throttle_cfg = &thread->throttle_cfg;
+
+	/* If the interval is very small, we do laps of 1 */
+	if (thread->workload->throttle < THROTTLE_OPS) {
+		throttle_cfg->tickets_per_increment = 1;
+		throttle_cfg->usecs_increment =
+			USEC_PER_SEC / thread->workload->throttle;
+	} else if (thread->workload->throttle < THROTTLE_INTVL) {
+		throttle_cfg->usecs_increment =
+		    USEC_PER_SEC / thread->workload->throttle * THROTTLE_OPS;
+		throttle_cfg->tickets_per_increment = THROTTLE_OPS;
+	} else {
+		/* If the interval is too large, we do more tickets per batch */
+		throttle_cfg->usecs_increment = THROTTLE_OPS;
+		throttle_cfg->tickets_per_increment =
+		    thread->workload->throttle / THROTTLE_INTVL;
```

Comment:
AFAICT, the code doesn't match the comment.  With 100000, usecs gets set to 100 and ticket ends up as 10.  I think that what we want to do here is reduce to two situations.  The first is the "very small" case.  The other is keeping the number of operations constant, and adjusting the usecs value.  So:
```
if (thread->workload->throttle < THROTTLE_OPS) {
    /* For very small throttle values, do one operation less frequently. */
    throttle_cfg->tickets_per_increment = 1;
    throttle_cfg->usecs_increment =
        USEC_PER_SEC / thread->workload->throttle;
else {
    /* Do THROTTLE_OPS each time, and adjust the number of times we do them, per second */
    throttle_cfg->tickets_per_increment = THROTTLE_OPS;
    throttle_cfg->usecs_increment =
        USEC_PER_SEC * THROTTLE_OPS  / thread->workload->throttle;
}
```
Hmm, this falls down if the throttle is set to more than 100million.  If we set THROTTLE_OPS to 1000, then we have a window of 1 billion ops per second.  We're probably safe there.  In any case, I think the choices are doing fixing the number of ops per period and adjusting the period, or fixing the period and adjusting the number of ops.  I like adjusting the period because it spreads out the scheduling and sleeping.